### PR TITLE
Add Login With Amazon auth provider

### PR DESCRIPTION
**Pull request by zendes**

### CODE REVIEW

**Comment by nicolai86** [comment]
when merged, this will fix #1

**Comment by nicolai86** [comment]
wd auth still works - I can't confirm amazon auth, but I assume it's working. pulling this in optimistically.

File: session.go [hunk 1]
```diff
@@ -190,43 +190,83 @@ func (c *Client) StartSession(ctx context.Context, username, password string) er
...
-	client_id := ""
-	redirect_url := ""
-	referer := ""
-	warpdrive_url := ""
-	{
+	requestURI := ""
+	deviceID := c.newID()
+	token := ""
+	switch data.Provider {
+	case "amazon":
```
Comment (harrisonhjones) [inline session.go]: Suggestion: lets move the functions that handle each provider to their own function. I think it'll make reviewing this diff 100% easier.

If we did that we could even do something like:

```go
loginProviders = map[string]AuthProvider{
    "amazon": amazonProvider,
    "wd": warpDriveProvider,
}
```

and then here we could do something like:


```go
authProvider, ok := authProviders[data.Provider]
if !ok {
    return fmt.Errorf("auth provider %q not yet supported", data.Provider)
}
authProvider.authenticate(...)
```

File: session.go [hunk 2]
```diff
@@ -370,7 +331,191 @@ func (c *Client) StartSession(ctx context.Context, username, password string) er
 		}
 		f(doc)
 
+		if requestURI == "" {
+			return fmt.Errorf("authentication for %s failed", username)
+		}
+
 		token = strings.Split(requestURI, "?")[1]
+
+	case "wd":
```
Comment (harrisonhjones) [inline session.go]: nit: can we add a comment? Something like `// Handle WarpDrive (Active Directory) authentication`

File: session.go [hunk 3]
```diff
@@ -190,43 +190,83 @@ func (c *Client) StartSession(ctx context.Context, username, password string) er
...
 		})
 		if err != nil {
 			return err
 		}
 
-		vs := resp.Request.URL.Query()
-		client_id = vs.Get("client_id")
-		redirect_url = vs.Get("redirect_uri")
-		referer = resp.Request.URL.String()
+		var appActionToken string
```
Comment (harrisonhjones) [inline session.go]: these lines could be replaced with the

```go
var x, y, z, ... string
```

form which is, imo, easier to read.

File: session.go [hunk 4]
```diff
@@ -190,43 +190,83 @@ func (c *Client) StartSession(ctx context.Context, username, password string) er
...
-	}
-
-	// Extract some OAuth stuff.
-	client_id := ""
-	redirect_url := ""
-	referer := ""
-	warpdrive_url := ""
-	{
+	requestURI := ""
+	deviceID := c.newID()
```
Comment (harrisonhjones) [inline session.go]: out of scope but this function now sufficiently large that we should rename `c` to something more descriptive.

File: session.go [hunk 5]
```diff
@@ -236,126 +276,47 @@ func (c *Client) StartSession(ctx context.Context, username, password string) er
...
-			body:            authPayload,
-			responseHandler: &responseHandler,
+			body:   form.Encode(),
+			output: &buf,
 		})
 		if err != nil {
 			return err
 		}
+		vs := resp.Request.URL.Query()
+		code := vs.Get("code")
```
Comment (harrisonhjones) [inline session.go]: You don't use code, scope, or state anywhere else so why not just:

```go
q.Set("code", vs.Get("code"))
...
```

File: session.go [hunk 6]
```diff
@@ -236,126 +276,47 @@ func (c *Client) StartSession(ctx context.Context, username, password string) er
...
-			},
-			body:            authPayload,
-			responseHandler: &responseHandler,
+			body:   form.Encode(),
+			output: &buf,
 		})
 		if err != nil {
 			return err
 		}
+		vs := resp.Request.URL.Query()
```
Comment (harrisonhjones) [inline session.go]: what does `vs` mean?

File: session.go [hunk 7]
```diff
@@ -370,7 +331,191 @@ func (c *Client) StartSession(ctx context.Context, username, password string) er
 		}
 		f(doc)
 
+		if requestURI == "" {
+			return fmt.Errorf("authentication for %s failed", username)
```
Comment (harrisonhjones) [inline session.go]: Can we not add any more description as to why here? Do we not know why at this point?

**Pull request merged**